src/DominantJuice.sol

```diff
@@ -0,0 +1,396 @@
+
+        if (address(this).balance == 0) {
+            hasCreatorWithdrawnAllFunds = true;
+        }
+    }
+
+    /// Getters and Helpers
+
+    // Due to decimnal precision and calculating the power of floating-point numbers, we use PRBMath for this helper function.
+    function calculateRefundBonus(address _pledger) public returns (uint256) {
+        require(pledgedAmount[_pledger] > 0, "Address is not a pledger.");
+
+        uint256 numPledgers = pledgers.length;
+        UD60x18 individualRefundBonus;
+        UD60x18 pledgeAmount_orderLoop;
+        UD60x18 pledgeAmount_amountLoop;
+
+        // Donations and order of pledges are weighted to benefit earlier, larger pledges
+        UD60x18 orderWeight;
+        UD60x18 sumOfOrderWeights;
+        // Memory array of normalized order weights (w-1s)
+        uint256[] memory orderWeightNormalized = new uint[](numPledgers);
+        UD60x18 sumOfAmountsXNormalizedOrderWeights;
+        // uint256 donationWeight;
+        // uint256 donationWeightNormalized;
+        // Represents the combined order and pledge amount weighting.
+        UD60x18 overallPledgeWeight;
+        emit Test("totalAmountPledged", totalAmountPledged);
+        // First, get the sum of pledge order weights to use in the normalizing loop
+        for (uint32 i = 0; i < numPledgers; i++) {
+            // Add 1 to i because rateOfDecay is being raised to the pledge ordereth power
+            sumOfOrderWeights = add(sumOfOrderWeights, rateOfDecay.powu(i + 1));
+        }
+
+        // Refund Bonus calculation prep work:
+        for (uint32 j = 0; j < numPledgers; j++) {
+            // Earlier pledges are given more weight according to the exponential decay rate.
+            // Add 1 to j because rateOfDecay is being raised to the ordereth power.
+            orderWeight = rateOfDecay.powu(j + 1);
+            // Order weights are normalized to compare each pledge to the total pledging data.
+            // Save in memory array for use in next for loop.
+            orderWeightNormalized[j] = unwrap(orderWeight / sumOfOrderWeights);
+            // Get the payment amount for each order
+            pledgeAmount_orderLoop = wrap(orderToPledgerToAmount[j][pledgers[j]]);
+            // Multiply payments by their respective normalized weight. Sum all products by
+            // looping (=sum(x*w_1)). Use in next for loop
+            sumOfAmountsXNormalizedOrderWeights =
+                add(sumOfAmountsXNormalizedOrderWeights, mul(pledgeAmount_orderLoop, wrap(orderWeightNormalized[j])));
+        }
+
+        // Next, give more weight to larger contribution amounts by multiplying each pledge amount with its
+        // orderWeightNormalized. Looping logic is for the possibility of repeat pledgers
+        for (uint32 k = 0; k < numPledgers; k++) {
+            if (pledgers[k] == _pledger) {
+                pledgeAmount_amountLoop = wrap(orderToPledgerToAmount[k][_pledger]);
```

Comment:
`k` runs from `0` to `numPledgers`. This line assumes that num orders is the same as num pledgers